source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+- developer tools: testing, debugging, linting and formatting, ...
+- generic build mechanisms: `stdenv`, trivial builders, ...
+- composition and configuration mechanisms: `override`, `overrideAttrs`, overlays, `callPackage`, ...
+- ecosystem-specific packaging mechanisms: `buildGoModule`, `buildPythonApplication`, ...
+- NixOS module system: `config`, `option`, ...
+
+**This guide only covers language syntax and semantics**, briefly discusses function libraries, and at the end will direct you to resources on the other components.
+
+### What will you learn?
+
+This guide should enable you to read typical Nix language code and understand its structure.
+
+It shows the most common and distingushing patterns in the Nix language:
+
+- [assigning names and accessing values](names-values)
+- declaring and calling [functions](functions)
+- [built-in and library functions](libraries)
+- [side effects](side-effects) to obtain build inputs and produce build results
+
+It *does not* explain all Nix language features in detail.
+See the [Nix manual][manual-language] for a full language reference.
+
+[manual-language]: https://nixos.org/manual/nix/stable/expressions/expression-language.html
+
+### What do you need?
+
+- Familiarity with software development
+- Familiarity with Unix shell, to read command line examples <!-- TODO: link to yet-to-be instructions on "how to read command line examples" -->
+- [](./install-nix), to run the examples
+
+### How long does it take?
+
+- **No experience with functional programming: 2 hours**
+
+  Carefully work through all definitions and examples.
+
+  Make sure you understand them by reading all detailed explanations.
+
+- **Familiar with functional programming: 1 hour**
+
+  Learn syntax and semantics by examining definitions and examples, and read detailed explanations if in doubt.
+
+- **Proficient with functional programming: 30 minutes**
+
+  The Nix language is largely unsurprizing to experienced functional programmers.
+
+  Read through the guide to get accustomed to the syntax and language features.
+
+We recommend running all examples and playing with them to validate your assumptions and test what you have learned.
+
+### How to run the examples?
+
+- Every valid piece of Nix language code is a *Nix expression*.
+- Nix expressions can contain other Nix expressions, that is, they can be nested.
+- Evaluating a Nix expression produces a single value.
+- Every *Nix file* (`.nix`) contains a single Nix expression.
+
+:::{note}
+To *evaluate* means to transform an expression according to the language rules until no further simplification is possible.
+:::
+
+All examples in this guide are valid Nix expressions.
+Each one is followed by the expected evaluation result.
+
+The following example is a Nix expression adding two numbers:
+
+```nix
+1 + 2
+```
+
+    3
+
+#### Interactive evaluation
+
+Use [`nix repl`][nix-repl] to evaluate Nix expressions interactively (by typing them on the command line):
```

Comment:
You need to mention `:p` for strict evaluation somewhere, it's needed for many examples

Reply:
Not strictly needed, but yes, sometimes otherwise the output will not be the same.